---
title: "Figures"
output:
  html_document:
    fig_height: 7
    fig_width: 9
    theme: cosmo
    toc: yes
  html_notebook:
    toc: yes
editor_options:
  chunk_output_type: inline
---
```{r}
basepath <- '.'
```

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_knit$set(root.dir=basepath, dpi=600, cache=TRUE)
knitr::opts_chunk$set(root.dir = basepath, dpi=600, cache=TRUE)

library(ComplexUpset)
library(data.table)
library(DESeq2)
library(easyalluvial)
library(EnhancedVolcano)
library(janitor)
library(ggalluvial)
library(ggsci)
library(ggthemes)
library(ggfortify)
library(ggpubr)
library(ggrepel)
library(gghighlight)
library(ggraph)
library(ggmosaic)
library(gwascat)
library(kableExtra)
library(igraph)
library(rstatix)
library(reticulate)
library(scales)
library(splitstackshape)
library(tidyverse)
library(tidygraph)
library(uwot)

reticulate::use_virtualenv(file.path(basepath, '.venv/'), required = T)
select <- dplyr::select

studycolors = c(
  '#4669bd', 
  'CD' = '#c25572',
  'UC' = '#ff9706',
  'nonIBD' = '#81c404',
  'ChEMBL'="#3c8c94", #Chembl
   'Enrichment analysis'='#f4e218', #Enrichment
   'GWAS'='#fb8072',
   'Negative' = 'firebrick4',
   'Positive' = 'steelblue', 
   'other' = 'black'
)
```

# Metabolomics 

```{r}
load(file.path(basepath,'tmp/metabolomics.RData'))

#Lookup tables
hmdb_ref <- fread(file.path(basepath,'tmp/hmdb_ref_w_origin.csv'))

#Load results
metab_results <- fread(file.path(basepath,'results/summaries/master_mbx.tsv'))
HMP2_metabs <- 
  readxl::read_xlsx(file.path(basepath,'tmp/Supplementary Tables 1-28.xlsx'), 
                    sheet=2) %>%
  left_join(metab_results %>% 
              select(Metabolites, HMP2_metab), 
            by = c('Feature'='HMP2_metab')) %>% 
  distinct_all

```

## UMAP 

```{r}

samumap <- metabolomics_raw %>% 
  mutate_if(is.numeric, ~replace_na(0)) %>% 
  select(8:533) %>% 
  t %>% 
  umap(n_neighbors = 50)


samples.umap <- metabolomics_raw %>% 
  select(8:533) %>% 
  names %>% 
  data.frame(external_id=.) %>% 
  left_join(sample_data %>% 
              filter(data_type=='metabolomics') %>% 
              select(external_id, diagnosis)) %>% 
  cbind(as.data.frame(samumap))%>% 
  ggplot() +
  aes(x=V1, y=V2, color=diagnosis, alpha=0.4) +
  geom_point()+
  theme_minimal()+ 
  theme(panel.border=element_blank(), 
        legend.background=element_blank())+
  guides(alpha='none')+
  scale_color_manual(values = studycolors) + 
  xlab('UMAP 1') + ylab('UMAP 2')
```
## Scatterplots
```{r}
consensus.mwu <- metab_results %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, MWU_delta, diagnosis) %>% 
  ggplot(aes(x=MWU_delta, y=ranking, color=diagnosis, label=Metabolites))+
  geom_jitter(aes(group=diagnosis), position=position_jitterdodge(), alpha=0.7) +
  facet_grid(.~diagnosis) +
  geom_text_repel(data = metab_results %>% 
                    distinct(Metabolites, diagnosis, MWU_delta, ranking) %>%
                    group_by(diagnosis) %>% 
                    top_n(10),
                  box.padding = 1,
                  show.legend = FALSE)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank())+
  ylab('Consensus score')+
  xlab('Differential abundance in disease compared to controls')+
  scale_color_manual(values = studycolors)
```

## Upsets
```{r, fig.width=12, fig.height=8}
up.df <- HMP2_metabs %>% 
  clean_names %>% 
  pivot_longer(cols=3:10, 
               names_to = c(".value", 'diagnosis'), 
               names_pattern = "(.+)_(.+)") %>% 
  mutate(diagnosis=str_to_upper(diagnosis)) %>% 
  select(-metabolites) %>% 
  full_join(metab_results, 
            by=c('feature'='HMP2_metab', 'diagnosis'))  %>% 
  select(-KEGG, -hgncid, -SMILES)%>% 
  distinct(feature, diagnosis, .keep_all=T) %>% 
  group_by(feature, diagnosis) %>%
  summarize(HMDB=HMDB, 
            ranking=ranking, 
            in_HMP2 = case_when(q_value<0.05~TRUE, TRUE~FALSE), 
            in_this_study = case_when(ranking>quantile(metab_results$ranking,.75)~TRUE, TRUE~FALSE)) %>% 
  ungroup %>% 
  left_join(hmdb_ref %>% 
              select(HMDB, origin)) %>% 
  mutate(origin = str_to_lower(origin),
         origin = case_when(
           feature == '2-hydroxyhexadecanoate' ~ 'endogenous, animal',
           feature == 'alpha-ketoglutarate' ~ 'endogenous, animal, plant, fungi, microbe',
           feature == 'carnosol(.)isomer' ~ 'plant', 
           feature == 'ectoine' ~ 'microbe',
           feature == 'diacetylspermine' ~ 'endogenous',
           origin == 'not available' | origin=='' | is.na(origin) ~ 'other', 
           TRUE~origin)
         ) %>%  
  cSplit('origin',', ','long') %>% 
  dcast(...~origin) %>% 
  distinct_all() %>% 
  mutate_at(vars(7:12), ~.!=0)

accessory.df <-  up.df %>% 
  filter(in_this_study) %>% 
  upset_data(colnames(up.df)[7:12]) %>% 
  `[[`('with_sizes')


met.upset.plot <- upset(data = up.df ,
      intersect = colnames(up.df)[7:12], 
      width_ratio = 0.2,
      keep_empty_groups=F,
      name = 'Metabolite sources intersections',
      themes = upset_modify_themes(
        list(
          'overall_sizes'=theme(axis.text.x=element_text(angle=90),
                                legend.position = 'none')
          )
        ),
      base_annotations=list(
        'Intersection size'= upset_annotate(
          y = '..count..',
          geom = list(
            geom_bar(data=accessory.df,
              aes(fill=diagnosis),
              position=position_dodge()),
            scale_fill_manual(values=studycolors, guide=FALSE)
            )
          )
        ),
       set_sizes=list(
         geom_point(data =up.df %>%
                      select(1, 7:12) %>% 
                      distinct_all() %>% 
                      upset_data(colnames(.)[2:7]) %>% 
                      `[` (c('intersected','matrix_frame')) %>% 
                      purrr::reduce(left_join) %>% 
                      filter(value) %>% 
                      group_by(group) %>% 
                      summarize(n_mets = n_distinct(feature)) %>% 
                      ungroup, 
                    aes(y=n_mets)),
         ylab('Metabolites by source'),
         scale_y_reverse()
        ),
      annotations = list(
        'Consensus scoring'= list(
          aes = aes(x=intersection, y = ranking, color=diagnosis),
          geom = list(geom_jitter(),
                      geom_boxplot(alpha=0.7),
                      scale_color_manual(values=studycolors, guide=FALSE),
                      # stat_compare_means(label = 'p.adj',
                      #                    hide.ns = T)
                      stat_pvalue_manual(data = up.df %>% 
                                           upset_data(colnames(up.df)[7:12]) %>% 
                                           `[[`('with_sizes')%>% 
                                           group_by(intersection) %>%
                                           filter(n_distinct(ranking)>2) %>% 
                                           wilcox_test(ranking~diagnosis,
                                                       detailed = T)%>%
                                           adjust_pvalue(method = 'fdr') %>% 
                                           add_significance('p.adj'), 
                                         label = "p.adj", 
                                         y.position = 1.2)
                      )
          )
        )
      ) 

```

## Fig1
```{r, fig.width=10, fig.height=8}
fig1top<- samples.umap + 
  ylab('') + 
  theme_minimal() + 
  consensus.mwu +
  theme_minimal() +
  guides(color=FALSE) +
  patchwork::plot_layout(widths = c(1.1,2))

fig1top/met.upset.plot + 
  patchwork::plot_layout(heights=c(1,2))+
  patchwork::plot_annotation(tag_levels = 'a')

ggsave(file.path(basepath,'results/manuscript_files/Fig1.pdf'), device = 'pdf', width = 16, height = 12, units = 'in')
```

## Load results
```{r}
#Load results
metab_results <- fread(file.path(basepath,'results/summaries/master_mbx.tsv')) 
htx_results <- 
  fread(file.path(basepath,'results/summaries/RNASeq_summaries.tsv')) %>%
  select(hgncid, contains('_CD')) %>%
  rename_at(vars(contains('_CD')), ~str_remove(.,'_CD'))%>%
  mutate(diagnosis='CD') %>%
  rbind(fread(file.path(basepath,'results/summaries/RNASeq_summaries.tsv')) %>% 
  select(hgncid, contains('_UC')) %>% 
  rename_at(vars(contains('_UC')), ~str_remove(.,'_UC'))%>% 
  mutate(diagnosis='UC'))

chembl_similars <- fread(file.path(basepath,'elaborated_data/CHEMBL25.CFP.csv.gz'))%>% 
  clean_names() %>% 
  select(HMDB = tar_name, analog=chembl_id, similarity) %>% 
  mutate(similarity_type = 'Tanimoto') %>% 
  bind_rows(fread(file.path(basepath,'elaborated_data/CHEMBL25.CFP_Tv05.csv.gz'))%>%
              clean_names() %>%
              select(HMDB = tar_name, analog=chembl_id, similarity) %>%
              mutate(similarity_type = 'Tversky_sub')
            )%>%
  bind_rows(fread(file.path(basepath,'elaborated_data/BioMAP_CHEMBL25.CFP.csv')) %>%
              clean_names() %>% 
              select(HMDB = tar_name, analog=chembl_id, similarity) %>% 
              mutate(similarity_type = 'Tanimoto')
            ) %>%
  mutate(association='ChEMBL')


chembl_assays<-fread(file.path(basepath,'elaborated_data/Chembl_f_assays_BIOME.csv')) %>% 
  # bind_rows(fread(file.path(basepath,'elaborated_data/Chembl_assays_BioMAP_BIOME.csv'))) %>% 
  mutate(association='ChEMBL') %>% 
  drop_na() %>% 
  mutate(type = case_when(
    grepl('EC|AC|activation|\\bagonist\\b', standard_type, ignore.case=T) ~ 'Positive',
    grepl('IC|Ki|inhibition|antagonist', standard_type, ignore.case=T) ~ 'Negative',
    # grepl('antagonist', assay_description, ignore.case = T) ~ 'Antagonist',
    # grepl('\\b[A,a]gonist\\b',assay_description) ~ 'Agonist',
    is.na(standard_type) ~ NA_character_,
    TRUE ~ 'other')) %>% 
  select(analog=compound_chembl_id, 
         pxc50 = pchembl_value,
         type,
         hgncid) %>%
  distinct_all


target_prioritization <- metab_results %>% 
  mutate(Metabolites = coalesce(Metabolites, HMP2_metab)) %>% 
  select(-hgncid) %>% 
  distinct_all %>% 
  left_join(chembl_similars) %>% 
  select(Metabolites, HMDB, ranking, MWU_delta, diagnosis, 
         direct_parent, super_class, analog, association, 
         similarity, similarity_type) %>% 
  left_join(chembl_assays) %>% 
  drop_na(Metabolites, HMDB, ranking, analog, similarity, hgncid) %>% 
  distinct_all %>% 
  drop_na(HMDB, hgncid) %>% 
  left_join(htx_results, by = c('hgncid', 'diagnosis')) %>%
  distinct_all

```

## Connect with targets
```{r}
t_classes <- fread(file.path(basepath, 'tmp/Target_annotation.csv')) %>% 
  readr::type_convert()
hmdb_ref <- fread(file.path(basepath,'tmp/hmdb_ref_w_origin.csv'))


HMP2_important_metabs <- 
  readxl::read_xlsx(file.path(basepath,'tmp/Supplementary Tables 1-28.xlsx'), 
                    sheet=2) %>%
  left_join(metab_results %>% 
              select(Metabolites, HMP2_metab), 
            by = c('Feature'='HMP2_metab')) %>% 
  distinct_all %>% 
  clean_names() %>% 
  filter(minimum_q_value<0.05) 

pri.met <- target_prioritization %>%  
  filter(ranking > quantile(ranking, .75))  %>% 
  filter(pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', 
                similarity > 0.85, 
                similarity >0.95)) %>% 
  group_by(hgncid) %>%
  mutate(
    mHits = n_distinct(HMDB)
    ) %>%
  mutate(
    Metab_hits = paste0(
      unique(Metabolites),
      collapse = "; ")
    ) %>%
  ungroup %>%
  group_by(HMDB) %>%
  mutate(
    gHits = n_distinct(hgncid)
    ) %>%
  mutate(
    gene_hits = paste0(
      unique(hgncid),
      collapse = "; ")
    ) %>%
  ungroup %>% 
  filter(gHits < 16) %>%
  arrange(desc(ranking)) %>%
  distinct_all %>% 
  group_by(Metabolites, hgncid) %>% 
  filter(pxc50==max(pxc50)) %>% 
  ungroup %>% 
  left_join(hmdb_ref %>% 
              select(Metabolites, super_class)) %>% 
  left_join(t_classes %>% 
              select(hgncid, targetclass)) %>% 
  distinct_all %>% 
  group_by(super_class) %>% 
  mutate(n_mets = n_distinct(Metabolites), 
         n_conn = n_distinct(targetclass))%>% 
  group_by(targetclass) %>% 
  mutate(n_targets = n_distinct(hgncid)) %>% 
  ungroup


```
## Fig 2
```{r}
mets.hist <- pri.met %>% 
  group_by(super_class) %>% 
  mutate(n_mets=n_distinct(Metabolites)) %>% 
  group_by(super_class, type, targetclass) %>% 
  summarize(n_targets=length(unique(hgncid)), 
            n_mets = max(n_mets)) %>% 
  ggplot()+
  aes(x=fct_rev(super_class),
      y=n_targets, 
      fill=targetclass)+
  geom_bar(stat='identity')+
  geom_text(data = pri.met %>% 
              group_by(super_class, type) %>% 
              summarize(n_mets=n_distinct(Metabolites)),
              aes(x=super_class,
                  y=-1, 
                  label=n_mets), 
            inherit.aes = F, 
            size=3)+
  coord_flip()+
  facet_grid(type~., scales = 'free', space = 'free')+
  theme_minimal()+
  scale_fill_tableau('Tableau 20', na.value='grey')+
  guides(fill=guide_legend(ncol=1, title = 'Target classes'))+
  theme(legend.position = 'left')+
  labs(x='Metabolite class', y = 'Number of connected targets per interaction type')
  
met.alluv <-  pri.met %>% 
  distinct(Metabolites, hgncid, .keep_all = T) %>% 
  group_by(Metabolites, targetclass) %>% 
  mutate(nconn = n_distinct(hgncid)) %>% 
  group_by(targetclass) %>% 
  mutate(n_targs=n_distinct(hgncid)) %>% 
  ungroup %>% 
  arrange(nconn) %>% 
  mutate(super_class=factor(super_class), 
         type=factor(type),
         targetclass=factor(targetclass)) %>% 
  select(super_class, type, targetclass) %>% 
  alluvial_wide(stratum_label_size = 3,
                order_levels = c('other','Positive','Negative'), 
                col_vector_flow = c(tableau_color_pal('Tableau 20')(length(unique(pri.met$targetclass))-1), 
                                    'grey'), 
                stratum_labels = F, 
                fill_by = 'last_variable') +
  geom_text_repel(aes(label = case_when(
    value=='Enzyme_all_others'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Enzyme_all_others',]$n_targets))),
    value=='7TM_Group1'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='7TM_Group1',]$n_targets))),
    value=='Other'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Other',]$n_targets))),
    value=='Transcriptional_Factor_all_others'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Transcriptional_Factor_all_others',]$n_targets))),
    value=='Nuclear Receptor'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Nuclear Receptor',]$n_targets))),
    value=='Kinase_Protein'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Kinase_Protein',]$n_targets))),
    value=='Ion Channel'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Ion Channel',]$n_targets))),
    value=='Transporter'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Transporter',]$n_targets))),
    value=='Protease'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Protease',]$n_targets))),
    # value=='7TM_all_others'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='7TM_all_others',]$n_targets))),
    # value=='Receptor_all_others'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Receptor_all_others',]$n_targets))),
    # value=='Enzyme_Esterase'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Enzyme_Esterase',]$n_targets))),
    # value=='Enzyme_Transferase'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Enzyme_Transferase',]$n_targets))),
    # value=='Extracellular Ligand'~ as.character(na.omit(unique(pri.met[pri.met$targetclass=='Extracellular Ligand',]$n_targets))),
    is.na(value)~ as.character(unique(pri.met[is.na(pri.met$targetclass),]$n_targets)),
    )),
    stat = "stratum", 
    direction = "y", 
    size=3,
    nudge_x = .2,
    color='#141414')+
  geom_text_repel(aes(label= ifelse(..stratum.. %in% pri.met$super_class, as.character(..stratum..), NA)), 
                  stat = "stratum", 
                  direction = "y", 
                  size=3,
                  nudge_x = -1,
             color='#141414')+
  geom_label(aes(label= ifelse(..stratum.. %in% pri.met$type, as.character(..stratum..), NA)), 
             stat = "stratum",
             size=3,
             color='#141414')+
  theme_minimal()+
  scale_x_discrete(labels = c('Metabolite class','Interaction type','Target class'))+
  labs(y='Interactions')



ggarrange(
  ggarrange(mets.hist+guides(fill='none'),
            met.alluv, 
            ncol = 1, 
            labels = 'auto', 
            heights = c(2,3)),
  cowplot::get_legend(mets.hist),
  widths = c(8,1)
)

ggsave(file.path(basepath,'results/manuscript_files/Fig2.pdf'), device = 'pdf', width = 12, height = 9, units = 'in')
```

```{r}
save.image(file.path(basepath,'tmp/metabolomics_w_figures.RData'))
```

# Host Transcriptomics
```{r, include=FALSE}
basepath <- '/home/an958266/projects/HMP2_Share'
select <- dplyr::select

studycolors = c(
  '#4669bd', 
  'CD' = '#c25572',
  'UC' = '#ff9706',
  'nonIBD' = '#81c404',
  'ChEMBL'="#3c8c94", #Chembl
   'Enrichment analysis'='#f4e218', #Enrichment
   'GWAS'='#fb8072',
   'Negative' = 'firebrick4',
   'Positive' = 'steelblue', 
   'other' = 'black'
)
```

## Load data

```{r}
load(file.path(basepath,'tmp/host_transcriptomics.RData'))
```


## UMAP

```{r}
sam.umap <- t(cts) %>% 
  as.data.frame %>% 
  rownames_to_column('external_id') %>% 
  left_join(metadata %>% 
              rownames_to_column('external_id') %>% 
              select(external_id, diagnosis, biopsy_location))%>% 
  umap(n_neighbors = 50)

samples.umap.diag <- t(cts) %>% 
  as.data.frame %>%
  data.frame(external_id=rownames(.)) %>% 
  left_join(metadata %>% 
              rownames_to_column('external_id') %>% 
              select(external_id, diagnosis, biopsy_location)) %>% 
  cbind(as.data.frame(sam.umap))%>% 
  ggplot() +
  aes(x=V1, y=V2, color=diagnosis, alpha=0.8) +
  geom_point()+
  theme_minimal()+ 
  theme(panel.border=element_blank(), 
        legend.background=element_blank())+
  guides(alpha='none')+
  scale_color_manual(values = studycolors) + 
  xlab('UMAP 1') + ylab('UMAP 2')

samples.umap.biopsy <- t(cts) %>% 
  as.data.frame %>%
  data.frame(external_id=rownames(.)) %>% 
  left_join(metadata %>% 
              rownames_to_column('external_id') %>% 
              select(external_id, diagnosis, biopsy_location)) %>% 
  cbind(as.data.frame(sam.umap))%>% 
  ggplot() +
  aes(x=V1, y=V2, color=biopsy_location, alpha=0.8) +
  geom_point()+
  theme_minimal()+ 
  theme(panel.border=element_blank(), 
        legend.background=element_blank())+
  guides(alpha='none')+
  scale_color_tableau() + 
  xlab('UMAP 1') + ylab('UMAP 2')

```

## Volcano plots

```{r}
labels <- res1df %>% 
  top_n(-5, log2FoldChange)%>% 
  bind_rows(res1df %>% 
              top_n(10, log2FoldChange)) %>% 
  mutate(diagnosis='UC') %>% 
  bind_rows(res2df %>%
              top_n(-5, log2FoldChange)%>%
              bind_rows(res2df %>% 
                          top_n(10, log2FoldChange)) %>% 
              mutate(diagnosis='CD'))

htx.volcano <- as.data.frame(res1)%>% 
  mutate(diagnosis='UC') %>% 
  bind_rows(as.data.frame(res2) %>% 
  mutate(diagnosis='CD')) %>% 
  ggplot()+
  aes(x=log2FoldChange, 
      y = -log(padj), 
      color=diagnosis)+
  geom_point(alpha=.75)+
  gghighlight(-log(padj)>2,
              abs(log2FoldChange)>.5, 
              calculate_per_facet=T)+
  geom_text_repel(data = labels, 
                  aes(label=hgncid), 
                  box.padding = 1, 
                  size = 3)+
  facet_grid(.~diagnosis) +
  theme_minimal() + 
  xlab( bquote(~Log[2]~ 'FC'))+
  ylab(bquote(~-Log[10]~adjusted~italic(p)~'-value'))+
  scale_color_manual(values = studycolors)

```


## Htx vs Mbx 

```{r}
mbx.htx.data <- target_prioritization %>% 
  select(hgncid, Metabolites, MWU_delta, analog, ranking,
         pxc50, similarity, similarity_type, type,
         log2FoldChange, padj, Pathway, Fold_Enrichment, highest_p, 
         diagnosis) %>% 
  filter(ranking > quantile(ranking, .75) | !is.na(log2FoldChange),
         pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', similarity > 0.85, similarity >0.95)) %>% 
  group_by(hgncid) %>%
  mutate(
    mHits = n_distinct(Metabolites)
  ) %>%
  mutate(
    Metab_hits = paste0(
      unique(Metabolites),
      collapse = "; ")
  ) %>%
  ungroup %>%
  group_by(Metabolites) %>%
  mutate(
    gHits = n_distinct(hgncid)
  ) %>%
  mutate(
    gene_hits = paste0(
      unique(hgncid),
      collapse = "; ")
  ) %>%
  ungroup %>% 
  filter(gHits <= 20, 
         mHits <= 20) %>%
  arrange(desc(ranking)) %>%
  distinct_all %>% 
  rename(Targets=hgncid) %>% 
  group_by(Metabolites, Targets, type) %>% 
  filter(pxc50 == max(pxc50)) %>% glimpse

label_data <- 
  mbx.htx.data %>% 
  group_by(diagnosis) %>% 
  filter(#abs(MWU_delta) > quantile(abs(MWU_delta), .75, na.rm=T), 
         #abs(log2FoldChange) > quantile(abs(log2FoldChange), .75, na.rm=T),
    Targets %in% c('CXCR1', 'CXCR2', 'HCAR2', 'GPR119','CHRNB2','NOS2'),
    Metabolites %in% c('Nicotinic acid','Ibuprofen', '2-Hydroxyibuprofen',
                       'Hydrocarboxylic acid','Trigonelline', 'Linoleoyl ethanolamide',
                       'L-Acetylcarnitine','L-arginine')
  ) %>% 
  ungroup %>% 
  distinct(Targets, Metabolites, diagnosis, log2FoldChange, MWU_delta, type) %>% 
  mutate(labpos.mbx=case_when(diagnosis=='CD' ~ -2.5, diagnosis=='UC'~5),
         labpos.htx=case_when(diagnosis=='CD' ~ -1, diagnosis=='UC'~1)) %>% 
  distinct_all()

mbx.htx.plot <- mbx.htx.data%>% 
  ggplot()+
  aes(x=MWU_delta, y=log2FoldChange)+
  geom_point(aes(size=-log(padj), alpha=ranking, color=diagnosis), show.legend = TRUE)+ 
  theme_minimal() +
  scale_color_manual(values = studycolors)+
  geom_text_repel(data=label_data,
                  aes(label=Metabolites, y=labpos.mbx,
                      color=type),
                  angle=90, direction = 'x',
                  show.legend = F
                  )+
  geom_vline(data=label_data,
             mapping = aes(xintercept = MWU_delta, color=type),
             linetype="dotdash")+
  geom_text_repel(data=label_data,
                  aes(label=Targets, x=labpos.htx),
                  color='green4',
                  direction = 'y',show.legend = F)+
  geom_hline(data=label_data,
             mapping = aes(yintercept = log2FoldChange),
             linetype="dotdash", color='green4', size=.5)+
  geom_point(data=label_data,
             shape=13, color='brown4', size=8)+
  facet_grid(~diagnosis)+
  xlab('Metabolite differential abundance')+
  ylab('Target differential expression') +
  geom_vline(xintercept = 0)+
  geom_hline(yintercept = 0)+
  guides(alpha=guide_legend(title='Metabolite consensus scoring'),
         size=guide_legend(
           title = bquote(atop("Target significance",
                               "(Log"[10]~adjusted~italic(p)~'-value)'))
         )
         
  ) 

```

## Fig 3
```{r}
ggarrange(
  ggarrange(samples.umap.biopsy+
              theme(legend.position = 'left')+
              guides(color=guide_legend(ncol=1)), 
            htx.volcano, 
            labels='auto',
            widths = c(4,5)),
  # pwy.plot,
  mbx.htx.plot +
    theme(legend.position = 'left'),
  nrow = 2, 
  heights = c(2,4),
  labels = c('','c'))

ggsave(file.path(basepath,'results/manuscript_files/Fig3.pdf'), device = 'pdf', width = 12, height = 10, units = 'in')
```

# Genetic associations
```{r}
IBD_pathways <- fread(file.path(basepath,'elaborated_data/Graham_Xavier_2020.tsv')) %>% 
    cSplit('action',';','long') %>% 
    cSplit('hgncid',';','long') %>% 
    group_by(hgncid) %>% 
    summarize(action=paste(unique(action),collapse='; '), 
              source=paste(unique(source),collapse='; ')
              )

data<- utils::data
gwas_catalog <- makeCurrentGwascat()
  
ids = grep("inflammatory|crohn|colitis|monocyte|lymphocyte", gwas_catalog$`DISEASE/TRAIT`, ignore.case = T)
IBD_Gwases = gwas_catalog[ids] %>% 
  as_tibble() %>% 
  cSplit('MAPPED_GENE',', ','long') %>% 
  cSplit('MAPPED_GENE',' - ','long') %>% 
  dplyr::rename(hgncid=MAPPED_GENE)

omim.table <- fread(file.path(basepath,'elaborated_data/Omim.table.csv'))

```


```{r}
target_prioritization_ge <- omim.table%>% 
  distinct(hgncid) %>% 
  full_join(IBD_Gwases) %>% 
  full_join(IBD_pathways) %>% 
  inner_join(target_prioritization) %>%
  group_by(hgncid, Metabolites, diagnosis) %>% 
  mutate(candidates = 
           case_when(MWU_delta<0 & type=='Negative'~ 'Negative modulator (scarce)', 
                     MWU_delta<0 & type=='Positive'~ 'Positive modulator (scarce)', 
                     MWU_delta>0 & type=='Positive'~ 'Positive modulator (abundant)', 
                     MWU_delta>0 & type=='Negative'~ 'Negative modulator (abundant)', 
                     TRUE~'Unknown')) %>% 
  filter(candidates != 'Unknown') %>%
  group_by(hgncid) %>% 
  mutate(
    mHits = n_distinct(HMDB)
  ) %>%
  ungroup %>% 
  group_by(HMDB) %>% 
  mutate(
    gHits = n_distinct(hgncid)
  ) %>% 
  ungroup %>% 
  arrange(desc(ranking)) %>%
  distinct_all

metabolites <- target_prioritization_ge %>% 
  select(label=Metabolites, weight=MWU_delta, significance=ranking, diagnosis) %>% 
  mutate(weight = sign(weight)) %>% 
  distinct_all

analogs <- target_prioritization_ge %>% 
  distinct(analog, diagnosis) %>% 
  rename(label=analog)

genes <- target_prioritization_ge %>% 
  select(label=hgncid, 
         weight=log2FoldChange, 
         significance=padj, 
         diagnosis) %>% 
  mutate(weight = sign(weight), 
         significance = replace_na(atan(-log(significance)), 0)) %>% 
  distinct_all

pwy <- target_prioritization_ge %>% 
  select(label=Pathway, 
         weight=Fold_Enrichment, 
         significance = padj,
         diagnosis) %>% 
  mutate(weight = sign(weight), 
         significance = replace_na(atan(-log(significance)), 0)) %>% 
  distinct_all

dis.traits <- target_prioritization_ge %>%
  select(label=DISEASE.TRAIT, 
         significance = padj,
         diagnosis) %>%
  distinct_all

nodes <- 
  metabolites %>% 
  full_join(analogs) %>% 
  full_join(genes) %>% 
  full_join(pwy) %>% 
  full_join(dis.traits) %>% 
  rowid_to_column("id")

similarities <-
  target_prioritization_ge %>% 
  distinct(Metabolites, analog, association, 
           similarity, similarity_type) %>% 
  left_join(nodes, by = c("Metabolites" = "label")) %>% 
  rename(from = id) %>% 
  left_join(select(nodes, label, id), 
            by = c("analog" = "label")) %>% 
  rename(to = id) %>% 
  select(from, to, everything())

assays <- 
  target_prioritization_ge %>% 
  group_by(analog, hgncid) %>% 
  filter(pxc50 == max(pxc50)) %>% 
  ungroup %>% 
  distinct(analog, hgncid, pxc50, type)%>% 
  left_join(nodes, by = c("analog" = "label")) %>% 
  rename(from = id, 
         association = type)%>% 
  left_join(select(nodes, label, id), 
            by = c("hgncid" = "label")) %>% 
  rename(to = id) %>% 
  select(from, to, everything())

pwy_e <- 
  target_prioritization_ge %>% 
  distinct(hgncid, pathway_enrichment,Pathway)%>% 
  mutate(association = 'Enrichment analysis') %>% 
  left_join(nodes, by = c("hgncid" = "label")) %>% 
  rename(from = id)%>% 
  left_join(select(nodes, label, id), 
            by = c("Pathway" = "label")) %>% 
  rename(to = id) %>% 
  select(from, to, everything()) %>% 
  drop_na
  
dis.tr_e <- 
  target_prioritization_ge %>% 
  select(hgncid, P.VALUE, DISEASE.TRAIT) %>% 
  distinct_all%>% 
  mutate(association = 'GWAS') %>% 
  left_join(nodes, by = c("hgncid" = "label")) %>% 
  rename(from = id)%>% 
  left_join(select(nodes, label, id), 
            by = c("DISEASE.TRAIT" = "label")) %>% 
  rename(to = id) %>% 
  select(from, to, everything()) %>% 
  drop_na

m2a <- graph_from_data_frame(d = similarities, vertices = nodes, directed = FALSE) %>% as_tbl_graph()

a2g <- graph_from_data_frame(d = assays, vertices = nodes, directed = FALSE) %>% as_tbl_graph()

g2p <- graph_from_data_frame(d = dis.tr_e, vertices = nodes, directed = FALSE) %>% as_tbl_graph()

g2d <- graph_from_data_frame(d = pwy_e, vertices = nodes, directed = FALSE) %>% as_tbl_graph()
  
fg <- m2a %>% 
  graph_join(a2g)%>% 
  graph_join(g2p) %>% 
  graph_join(g2d) %>% 
  activate('edges') %>% 
  mutate(`Similarity or binding`= 
           case_when(similarity_type =='Tanimoto' & similarity < 0.8 ~ 'weak',
                     similarity_type=='Tanimoto' & similarity >= 0.8 & similarity < 0.9 ~ 'good', 
                     similarity_type=='Tanimoto' & similarity >= 0.9 ~ 'strong',
                     similarity_type =='Tversky_sub' & similarity < 0.9 ~ 'weak',
                     similarity_type=='Tversky_sub' & similarity >= 0.9 & similarity < 0.95 ~ 'good', 
                     similarity_type=='Tversky_sub' & similarity >= 0.95 ~ 'strong', 
                     pxc50 < 5.5 ~ 'weak',
                     pxc50 >= 5.5 & pxc50 < 7 ~ 'good',
                     pxc50 > 7 ~ 'strong',
                     TRUE~NA_character_)) %>% 
  mutate(`Similarity or binding`=factor(`Similarity or binding`, 
                                        levels = c(NA,'weak','good','strong'), 
                                        ordered = TRUE)) %>% 
  activate('nodes')

quicknet <- qgraph(fg, 
       #node_label = label,
       node_colour = weight,
       edge_colour = association) 

quicknet + facet_nodes(diagnosis~.)

```
## Fig 4

```{r, fig.width=14, fig.width=15}
test_genes <- target_prioritization_ge%>%
  distinct(hgncid) %>%
  pull

saveplots <- 'results/summaries/target_plots/'

dir.create(file.path(basepath,saveplots))

plotlist <- list()

create_graph <- function(i){
  
  gene_of_interest <- test_genes[i]

  subnet <- fg%>% 
    tidygraph::as.igraph() %>% 
    neighborhood(nodes = fg %>% 
                   filter(label ==gene_of_interest) %>% 
                   as.data.frame() %>% 
                   `$` (name) %>%  
                   as.numeric, 
                 order =2
    )
  if(length(subnet)>0){
  
   gg <- igraph::induced_subgraph(fg, vids = unlist(subnet)) %>% 
    tidygraph::as_tbl_graph() %>% 
    ggraph(layout = 'linear', circular = TRUE) +
    geom_node_point(aes(size=significance), 
                    repel = TRUE) +
    geom_edge_arc(aes(color=association, alpha=`Similarity or binding`)) + 
    scale_edge_color_manual(values = studycolors,
                            na.value = '#343642',
                            drop = FALSE) + #Network
    scale_edge_alpha_discrete(limits = c('weak','good','strong'))+
    geom_node_label(mapping = aes(label=label, 
                                  colour=weight), 
                    repel = TRUE,
                    box.padding = 1,
                    ) + 
    scale_color_gradient2_tableau(limits=c(-2,2)) +
    theme_graph(border = F,foreground = 'steelblue', fg_text_colour = 'white', base_family = 'Helvetica')+
    theme(legend.position = 'right', legend.text = element_text(size=10))
   
   
   if(n_distinct(gg$data[gg$data$label==gene_of_interest,]$diagnosis)>1){

     plotlist[[i]] <- gg + facet_nodes(diagnosis~.)
     
   }else{

     plotlist[[i]] <- gg + facet_edges(diagnosis~.)
     
   }

    # ggsave(paste0(file.path(basepath,saveplots), gene_of_interest, '.png'), 
    #        device = 'png', dpi = 720,height = 6,  width = 10)
    
  }else{
  cat('Cannot plot subnet for',  gene_of_interest, '\n')
    }
}

plotlist <- lapply(seq(length(test_genes)), create_graph)

plt1 <- plotlist[match('NOS2', test_genes)]
plt234<- plotlist[match(c('RARB','GRM4','SLC22A3'), test_genes)]

ggpubr::ggarrange(ggpubr::ggarrange(plotlist = plt1, 
                                    common.legend = T, 
                                    ncol=1, legend = 'none'),
                  ggpubr::ggarrange(plotlist = plt234, 
                                    labels=c('b','c','d'), 
                                    common.legend = T, 
                                    legend='none', 
                                    ncol=1, 
                                    heights = c(2,3,2)),
                  cowplot::get_legend(plt1[[1]]),
                  ncol = 3,
                  labels=c('a',''), 
                  widths = c(3,2.5, 1))

ggsave(file.path(basepath,'results/manuscript_files/Fig4.pdf'), device = 'pdf', width = 16, height = 10, units = 'in')

```

# BioMAP
## Fig 5
```{r BioMAP, fig.width=15, fig.height=9}
bioMAP_plot <- readxl::read_xlsx(file.path(basepath, 'results/manuscript_files/Supplementary_Table5.xlsx'),sheet = 1) %>% 
  group_by(metabolite, assay_name_required) %>%
  arrange(concentration_molar) %>% 
  mutate(conc_levels = order(concentration_molar)) %>%
  ungroup %>% 
  arrange(metabolite, conc_levels, bio_marker_name_s_required) %>% 
  ggplot()+
  aes(y=log_ratio_required, x=bio_marker_name_s_required)+
  facet_grid(metabolite~system_name_required, scales = 'free', space = 'free_x')+
  theme_minimal()+
  geom_ribbon(aes(ymin=-significance_prediction_envelope_at_95_percent, ymax=significance_prediction_envelope_at_95_percent), group=1, alpha=0.2) +
  geom_line(aes(color=factor(conc_levels), group=factor(conc_levels))) +
  scale_color_tableau('Nuriel Stone', 
                      name='Metabolite concentration level')+
  theme(axis.text.x = element_text(angle=90, size = 8, vjust = .8), 
        legend.position = 'bottom',
        panel.spacing = unit(0 , "lines"))+
  ylab('LogFC vs control')+
  xlab('Biomarkers')

bioMAP_plot
ggsave(file.path(basepath,'results/manuscript_files/Fig5.pdf'), device = 'pdf', width = 15, height = 10, units = 'in')
```

## Table 3

```{r}
readxl::read_xlsx(file.path(basepath, 'results/manuscript_files/Supplementary_Table5.xlsx'),sheet = 1) %>%
  group_by(metabolite, assay_name_required) %>%
  arrange(concentration_molar) %>% 
  mutate(conc_levels = order(concentration_molar)) %>%
  ungroup %>% 
  distinct(metabolite, concentration_molar, conc_levels) %>% 
  arrange(metabolite, desc(concentration_molar)) %>% 
  type_convert() %>% 
  mutate(concentration_molar=concentration_molar*1e6) %>% 
  dplyr::rename(Metabolites = metabolite, 
         `Concentration (µM)`=concentration_molar,
         Level = conc_levels) %>% 
  kable(format = 'html', escape=F, digits = 2) %>% 
  kable_styling(bootstrap_options = 'condensed') %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = c(1), valign = "middle", row_group_label_position = 'stack')

```

```{r}
save.image(file.path(basepath,'tmp/host_transcriptomics_w_figures.RData'))
rm(list = ls())
```

```{r}
sessionInfo()
```


