---
title: "Supplementary_tables"
author: "Andy Nuzzo, Human Genetics Computational Biology"
date: "`r Sys.Date()`"
output: 
  html_notebook: 
    fig_caption: yes
    theme: paper
    toc: yes
    toc_float: yes
  bookdown::word_document2:
    fig_caption: yes
    reference_docx: '../../Gepotidacin/reference.docx'
    fig_height: 6
    fig_width: 8
    highlight: tango
    toc: yes
header-includes:
  - \usepackage{helvet}
---
```{r, include=FALSE}
basepath <- '/home/an958266/projects/HMP2_Share'
```

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_knit$set(root.dir=basepath, dpi=600)
knitr::opts_chunk$set(root.dir = basepath, dpi=600)
options(knitr.kable.NA = '--')

library(arsenal)
library(biomaRt)
library(ComplexUpset)
library(data.table)
library(DESeq2)
library(EnhancedVolcano)
library(ggsci)
library(ggthemes)
library(ggfortify)
library(ggpubr)
library(ggrepel)
library(gghighlight)
library(ggraph)
library(ggmosaic)
library(glue)
library(gwascat)
library(htmltab)
library(igraph)
library(janitor)
library(kableExtra)
library(rstatix)
library(reticulate)
library(scales)
library(splitstackshape)
library(tidygraph)
library(tidyverse)
library(uwot)

reticulate::use_virtualenv(file.path(basepath, '.venv/'), required = T)
select <- dplyr::select

studycolors = c(
  '#4669bd', 
  'CD' = '#c25572',
  'UC' = '#ff9706',
  'nonIBD' = '#81c404',
  'ChEMBL'="#3c8c94", #Chembl
   'Enrichment analysis'='#ffff33', #Enrichment
   'GWAS'='#fb8072',
   'Negative' = 'firebrick4',
   'Positive' = 'steelblue', 
   'other' = 'black'
)
```

# Load data
```{r}
load(file.path(basepath,'tmp/host_transcriptomics_w_figures.RData'))
```

# Metabolomics results

## Volcano plot
```{r}
mwu.volcano <- mwu%>% 
  select(HMDB, contains('pval')) %>%
  pivot_longer(-HMDB,
               names_to = 'diagnosis', 
               names_pattern = '.*(..)',
               values_to = 'pvalue_mwu', 
               values_drop_na = TRUE) %>% 
  full_join(mwu%>% 
              select(HMDB, contains('diff')) %>%
              pivot_longer(-HMDB,
                           names_to = 'diagnosis', 
                           names_pattern = '.*(..)',
                           values_to = 'delta', 
                           values_drop_na = TRUE)) %>% 
  distinct_all %>% 
  mutate(logq = -log(pvalue_mwu)) %>% 
  ggplot(aes(x=delta, y=logq, color=diagnosis))+
  geom_jitter(aes(group=diagnosis), position=position_jitterdodge(), alpha=0.7) +
  facet_grid(.~diagnosis) +
  theme_minimal()+
  ylab(bquote(~Scaled~significance~(-log(italic(q)~-value))))+
  xlab('Differential abundance in disease compared to controls')+
  scale_color_manual(values = studycolors)
  
```
## XGB
```{r}
xgb.cm <- cm$table %>% 
  t %>% 
  `[`(,ncol(.):1) %>% 
  melt %>% 
  ggplot(aes(x=Reference, y=Prediction, fill=value)) + 
  geom_tile()+
  theme_minimal()+
  scale_fill_gradient(low='white',high=studycolors[1])+
  xlab("True")+
  ylab("Predicted")+
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1),
        legend.position = 'bottom')+
  geom_text(aes(label = value), vjust = -.5)+
  geom_text(aes(label = percent(value/sum(value))), vjust = 1)

xgb.fi <- fread(file.path(basepath,'results/ML/mbx_XGB_fi.csv')) %>% 
  top_n(20) %>%
  ggplot(aes(x=reorder(features, percentage), 
             y=percentage))+
  geom_histogram(stat = 'identity', 
                 fill = studycolors[1])+
  ylab('Feature importance')+
  scale_y_continuous(labels = scales::percent_format(scale = 1e2, accuracy = .1))+
  xlab('')+
  coord_flip()+
  theme_minimal()

xgb.shap <- fread(file.path(basepath,'results/ML/mbx_XGB_shapi.csv')) %>%
  rowwise() %>% 
  mutate(wt = sum(nonIBD, UC, CD)) %>%
  ungroup %>% 
  top_n(n = 20) %>% 
  select(-wt) %>% 
  melt %>% 
  ggplot(aes(x=reorder(mbx, value),
             y=value, 
             color=variable, 
             fill=variable))+
  geom_bar(stat='identity') +
  coord_flip()+
  theme_minimal()+
  xlab('')+
  ylab('SHAP values')+
  scale_fill_manual(name = 'diagnosis', 
                    values=studycolors)+
  scale_color_manual(name = 'diagnosis',
                     values=studycolors)+
  theme(legend.position = 'right')

xgb.plot <- ggarrange(xgb.cm+
  theme(legend.position = 'none'), 
  xgb.fi, 
  xgb.shap+
  theme(legend.position = 'none'), 
  nrow = 1, 
  widths = c(2,3,4))
```

## GMML
### Cluster type
```{r}
gmml.cm <- cm2$table %>% 
  t %>% 
  `[`(,ncol(.):1) %>% 
  melt %>% 
  ggplot(aes(x=Reference, y=Prediction, fill=value)) + 
  geom_tile()+
  theme_minimal()+
  scale_fill_gradient(low='white',high=studycolors[1])+
  xlab("True")+
  ylab("Predicted")+
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1),
        legend.position = 'bottom')+
  geom_text(aes(label = value), vjust = -.5)+     
  geom_text(aes(label = percent(value/sum(value))), vjust = 1)

gmml.fi <- fread(file.path(basepath,'results/GMML/mbx_gmme_XGBClassifier_fi.csv'))  %>% 
  top_n(20) %>%
  ggplot(aes(x=reorder(features, percentage), 
             y=percentage))+
  geom_histogram(stat = 'identity', 
                 fill = studycolors[1])+
  ylab('Feature importance')+
  scale_y_continuous(labels = scales::percent_format(scale = 1e2, accuracy = .1))+
  xlab('')+
  coord_flip()+
  theme_minimal()

gmml.shap <- fread(file.path(basepath,'results/GMML/mbx_gmme_XGBClassifier_shapi.csv'))%>%
  rowwise() %>% 
  mutate(wt = sum(nonIBD, UC, CD)) %>%
  ungroup %>% 
  top_n(n = 20) %>% 
  select(-wt) %>% 
  melt %>% 
  ggplot(aes(x=reorder(mbx, value),
             y=value, 
             color=variable, 
             fill=variable))+
  geom_bar(stat='identity') +
  coord_flip()+
  theme_minimal()+
  xlab('')+
  ylab('SHAP values')+
  scale_fill_manual(name = 'diagnosis', 
                    values=studycolors)+
  scale_color_manual(name = 'diagnosis',
                     values=studycolors)+
  theme(legend.position = 'right')
```

### Mixed effects type
```{r}
gmml.me.cm <- cm3$table %>% 
  t %>% 
  `[`(,ncol(.):1) %>% 
  melt %>% 
  ggplot(aes(x=Reference, y=Prediction, fill=value)) + 
  geom_tile()+
  theme_minimal()+
  scale_fill_gradient(low='white',high=studycolors[1])+
  xlab("True")+
  ylab("Predicted")+
  theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1),
        legend.position = 'bottom')+
  geom_text(aes(label = value), vjust = -.5)+     
  geom_text(aes(label = percent(value/sum(value))), vjust = 1)

gmml.me.fi <- fread(file.path(basepath,'results/GMML/mbx_gmme_meXGBClassifier_fi.csv'))  %>% 
  top_n(20) %>%
  ggplot(aes(x=reorder(features, percentage), 
             y=percentage))+
  geom_histogram(stat = 'identity', 
                 fill = studycolors[1])+
  ylab('Feature importance')+
  scale_y_continuous(labels = scales::percent_format(scale = 1e2,accuracy = .1))+
  xlab('')+
  coord_flip()+
  theme_minimal()

gmml.me.shap <- fread(file.path(basepath,'results/GMML/mbx_gmme_me_XGBClassifier_shapi.csv'))%>%
  rowwise() %>% 
  mutate(wt = sum(nonIBD, UC, CD)) %>%
  ungroup %>% 
  top_n(n = 20) %>% 
  select(-wt) %>% 
  melt %>% 
  ggplot(aes(x=reorder(mbx, value),
             y=value, 
             color=variable, 
             fill=variable))+
  geom_bar(stat='identity') +
  coord_flip()+
  theme_minimal()+
  xlab('')+
  ylab('SHAP values')+
  scale_fill_manual(name = 'diagnosis', 
                    values=studycolors)+
  scale_color_manual(name = 'diagnosis',
                     values=studycolors)+
  theme(legend.position = 'right')
```

## GLL 
```{r}
gmml_mod_me <- py_load_object(file.path(basepath,'tmp/GMML/mbx_gmme_me_XGBClassifier.sav'))
gmml_mod_cl <- py_load_object(file.path(basepath,'tmp/GMML/mbx_gmme_cl_XGBClassifier.sav'))

gll.df <- gmml_mod_cl$gll_history %>% 
  as.matrix %>% 
  unlist %>% 
  as.data.frame() %>% 
  rename(gll=1) %>% 
  bind_cols(gmml_mod_cl$r2_history%>% 
              unlist %>% 
              as.matrix %>% 
              as.data.frame() %>% 
              rename(r2=1)) %>% 
  mutate(mod='clusters', 
         iterations = seq(1:20)) %>% 
  bind_rows(
    gmml_mod_me$gll_history %>% 
      as.matrix%>% 
      unlist %>% 
      as.data.frame() %>% 
      rename(gll=1) %>% 
      bind_cols(gmml_mod_me$r2_history %>% 
                  as.matrix%>% 
                  unlist %>% 
                  as.data.frame() %>% 
                  rename(r2=1)) %>% 
      mutate(mod='mixed_effects', 
             iterations = seq(1:20))
    ) %>% 
  pivot_longer(-c(iterations, mod), 
               names_to = 'diag')

gll.plot <- ggplot(gll.df)+
  aes(x=iterations, 
      y=value, 
      color = mod)+
  geom_line()+
  facet_wrap(diag~., scales = 'free_y')+
  theme_minimal()+
  scale_color_tableau(name='Generalized \nML model')+
  ylab('')+
  theme(legend.position='bottom')


```

## FigS1
```{r, fig.width=15, fig.height=8}
fig.s1 <- ggarrange(
  ggarrange(
    ggarrange(mwu.volcano+
                theme(legend.position = 'none'), 
              xgb.plot,
              ncol = 2,
              widths = c(2,5), 
              labels = 'auto'),
    ggarrange(gll.plot+
                theme(legend.position = 'none'),
              ggarrange(
                 ggarrange(gmml.cm+
                             theme(legend.position = 'none'), 
                           gmml.fi, 
                           gmml.shap+
                             theme(legend.position = 'none'),
                           nrow=1, 
                           widths = c(2,3,4)
                           ), 
                 ggarrange(gmml.me.cm+
                             theme(legend.position = 'none'), 
                           gmml.me.fi, 
                           gmml.me.shap+
                             theme(legend.position = 'none'),
                           nrow = 1,
                           widths = c(2,3,4)
                 ),
                 nrow=2, 
               labels = c('d','e')
               ),
               ncol = 2,
               widths = c(2,5), 
               labels = c('c','')),
    nrow = 2,
    heights = c(1,2)
    ),
  ggarrange(cowplot::get_legend(gll.plot), 
            cowplot::get_legend(gmml.cm), 
            cowplot::get_legend(gmml.me.shap+theme(legend.position = 'bottom')), 
            nrow = 1),
  ncol = 1, heights = c(10,1)
  )%>% show

ggsave(file.path(basepath,'results/manuscript_files/FigS1.pdf'), device = 'pdf', width = 19, height = 10, units = 'in')

```

## Supplementary Table 1
```{r}
fread(file.path(basepath,'results/summaries/mwu_summaries.csv')) %>% 
  full_join(fread(file.path(basepath,'results/summaries/XGB_res_summary.tsv'))) %>% 
  full_join(fread(file.path(basepath,'results/summaries/GMME_XGB_res_summary.tsv'))) %>% 
  select(-abundance) %>% 
  separate(HMDB, c('HMDB','HMP2_metab'), sep='; ') %>% 
  select(-HMDB) %>% 
  left_join(fread(file.path(basepath,'tmp/Manual_HMBD_ref.txt')) %>% 
              select(-SMILES), 
            by='HMP2_metab') %>%
  left_join(hmdb_ref_table)%>% 
  left_join(master_mbx %>% 
              distinct(HMDB, diagnosis, ranking) %>% 
              pivot_wider(names_from = diagnosis, 
                          values_from = ranking, 
                          names_prefix = 'scoring_',
                          values_fn = list(ranking = mean))
            ) %>% 
  select(HMDB, everything(), -KEGG, -hgncid) %>% 
  distinct_all() %>% 
  openxlsx::write.xlsx(file.path(basepath,'results/manuscript_files/Supplementary_Table1.xlsx'))

  
```


# Scoring evaluation
## Pre-transform
```{r}
importances.plot <- fread(file.path(basepath,'results/summaries/mwu_summaries.csv')) %>% 
  full_join(fread(file.path(basepath,'results/summaries/XGB_res_summary.tsv'))) %>% 
  full_join(fread(file.path(basepath,'results/summaries/GMME_XGB_res_summary.tsv'))) %>% 
  group_by(HMDB) %>%
  mutate_at(vars(contains('pvalue')), 
            list(~case_when(is.numeric(.) & .==0 ~ 0.1,
                            is.na(.) ~ 0.1,
                            is.numeric(.) ~ -log10(.), 
                            TRUE~.)
                 ))%>% 
  ungroup %>% 
  mutate_at(vars(contains('fi'), contains('shap'), 
                 'gain', contains('pvalue')), 
            list(~ case_when(is.numeric(.) ~ rescale(., to=c(0,1)),
                            TRUE~.))) %>%
  pivot_longer(cols=c(contains('pvalue'), contains('fi'),
                      contains('shap'), 'gain')) %>% 
  mutate(diagnosis = case_when(grepl('CD', name) ~'CD', 
                               grepl('UC', name) ~'UC',
                               grepl('nonIBD', name)~'nonIBD', 
                               TRUE~NA_character_)) %>% 
  mutate(name=str_remove_all(name, '_CD|_UC|_nonIBD')) %>% 
  ggplot(aes(x=name, y=value, color=diagnosis))+
  geom_jitter(aes(group=diagnosis), 
              position=position_jitterdodge(), 
              alpha=0.7) +
  theme_minimal() + 
  coord_flip() + 
  scale_color_manual(values = studycolors, 
                     na.value=studycolors[1])+
  scale_x_discrete(labels = c(~ atop(paste("MannWithney U"), 
                                     paste('log(', italic("q"), 
                                           '-value)')),
                              'Feature Importance \n(XGB)',
                              'Feature gain \n(XGB)',
                              'SHAP values \n(XGB)',
                              'Feature Importance \n(GMML with clusters)',
                              'SHAP values \n(GMML with \nrandom effects)'),
                   limits = c('pvalue_mann_whitney',
                              'xgb_fi','gain','shap_xgb',
                              'gmme_fi','shap_gmme'))+
  labs(x='', y='Value') 

```

## Consensus
```{r}
scoring.plot <- master_mbx %>% 
  group_by(HMDB) %>%
  distinct(HMDB, ranking, diagnosis) %>% 
  ggplot(aes(x=diagnosis, y=ranking, color=diagnosis))+
  geom_violin()+
  geom_jitter(aes(group=diagnosis), position=position_jitterdodge(), alpha=0.7) +
  theme_minimal() + 
  scale_color_manual(values = studycolors)+
  labs(y='Consensus Scoring', x='')+
  stat_compare_means(comparisons = list(c('CD','UC')), 
                     label = 'p.signif', 
                     p.adjust.method='BH')

```

#Scatterplots
```{r}
consensus.p_mwu <- 
  master_mbx %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, MWU_qval, diagnosis) %>% 
  ggplot(aes(x=MWU_qval, y=ranking, color=diagnosis))+
  geom_jitter(aes(group=diagnosis), position=position_jitterdodge(), alpha=0.7) +
  facet_grid(.~diagnosis) +
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  geom_smooth()+
  ylab('Consensus score')+
  xlab(~ atop(paste("MannWithney U"), 
              paste('log(', italic("q"),
                    '-value)')))+
  scale_color_manual(values = studycolors)
  
consensus.xgb_fi <- 
  master_mbx %>% 
  full_join(fread(file.path(basepath,'results/summaries/XGB_res_summary.tsv')) %>% 
              separate(HMDB, c('HMDB','HMP2_metab'), sep='; ')) %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, xgb_fi, diagnosis) %>% 
  drop_na() %>% 
  ggplot(aes(x=xgb_fi, y=ranking, color=diagnosis))+
  geom_jitter(alpha=0.7) +
  facet_grid(~diagnosis)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  geom_smooth()+
  ylab('Consensus score')+
  xlab('Feature Importance \n(XGB)')+
  scale_color_manual(values = studycolors)

consensus.xgb_gain <- master_mbx %>% 
  full_join(fread(file.path(basepath,'results/summaries/XGB_res_summary.tsv')) %>% 
              separate(HMDB, c('HMDB','HMP2_metab'), sep='; ')) %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, gain, diagnosis) %>% 
  drop_na() %>% 
  ggplot(aes(x=gain, y=ranking, color=diagnosis))+
  geom_jitter(alpha=0.7) +
  facet_grid(~diagnosis)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  geom_smooth()+
  ylab('Consensus score')+
  xlab('Feature gain \n(XGB)')+
  scale_color_manual(values = studycolors)

consensus.shap_xgb <-master_mbx %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, XGB_shap, diagnosis) %>% 
  drop_na() %>% 
  ggplot(aes(x=XGB_shap, y=ranking, color=diagnosis))+
  geom_jitter(alpha=0.7) +
  facet_grid(~diagnosis)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  geom_smooth()+
  ylab('Consensus score')+
  xlab('SHAP values \n(XGB)')+
  scale_color_manual(values = studycolors)
  
consensus.gmme_fi <- 
  master_mbx %>% 
  full_join(fread(file.path(basepath,'results/summaries/GMME_XGB_res_summary.tsv')) %>% 
              separate(HMDB, c('HMDB','HMP2_metab'), sep='; ')) %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, gmme_fi, diagnosis) %>% 
  drop_na() %>% 
  ggplot(aes(x=gmme_fi, y=ranking, color=diagnosis))+
  geom_jitter(alpha=0.7) +
  facet_grid(~diagnosis)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  geom_smooth()+
  ylab('Consensus score')+
  xlab('Feature Importance \n(GMML with clusters)')+
  scale_color_manual(values = studycolors)

consensus.shap_gmme <-master_mbx %>% 
  group_by(HMDB) %>%
  distinct(Metabolites, ranking, GMME_shap, diagnosis) %>% 
  drop_na() %>% 
  ggplot(aes(x=GMME_shap, y=ranking, color=diagnosis))+
  geom_jitter(alpha=0.7) +
  facet_grid(~diagnosis)+
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank(), 
        legend.position = 'none')+
  ylab('Consensus score')+
  xlab('SHAP values \n(GMML with \nrandom effects)')+
  scale_color_manual(values = studycolors)

```
## Qqplots
```{r}
consensus.qq <- master_mbx %>% 
  ggqqplot(x='ranking',
           color = 'diagnosis',
           alpha=.7)+
  facet_grid(.~diagnosis)+
  theme_minimal() + 
  scale_color_manual(values = studycolors)+
  scale_fill_manual(values = studycolors)+
  theme(legend.position = 'none')

```

## Fig S2
```{r}
fig.s2 <- ggarrange(
  ggarrange(importances.plot, scoring.plot, 
            nrow=1, labels='auto', 
            common.legend = T, legend = 'bottom'),
  ggarrange(
    consensus.p_mwu, consensus.xgb_fi, consensus.shap_xgb,
    consensus.xgb_gain, consensus.gmme_fi, consensus.shap_gmme,
    nrow = 2,
    ncol = 3,
    labels = c('c','d','e','f','g','h')
  ), 
  consensus.qq, 
  nrow = 3,
  heights = c(1,2,1),
  labels = c('','','i')
) %>% show

ggsave(file.path(basepath,'results/manuscript_files/FigS2.pdf'), device = 'pdf', width = 12, height = 14, units = 'in')

```


# HMP2 study comparison

## UMAP
```{r}
# metumap <- metabolomics_raw %>% 
#   select(8:533) %>% 
#   mutate_if(is.numeric, ~replace_na(0)) %>% 
#   umap(n_neighbors = 50, n_threads = 20)
# 
# metabolites.umap <- metabolomics_raw %>% 
#   select(1:7) %>% 
#   cbind(metumap) %>% 
#   ggplot() +
#   aes(x=V1, y=V2, color=Method, alpha=0.4) +
#   geom_point()+
#   theme_clean()+ 
# theme(panel.border=element_blank(), 
#       legend.background=element_blank())+
#   guides(alpha='none')+
#   scale_color_npg()


```

## Scatterplots
```{r}
consensus.hmp2.sig <- HMP2_metabs %>% 
  clean_names %>% 
  pivot_longer(cols=3:10, 
               names_to = c(".value", 'diagnosis'), 
               names_pattern = "(.+)_(.+)") %>% 
  mutate(diagnosis=str_to_upper(diagnosis)) %>% 
  select(-metabolites) %>% 
  full_join(metab_results, 
            by=c('feature'='HMP2_metab', 'diagnosis'))  %>% 
  select(-KEGG, -hgncid, -SMILES)%>% 
  distinct_all %>% 
  ggplot()+
  aes(x=q_value, y=ranking, color=diagnosis)+
  geom_jitter(alpha=0.7)+
  facet_grid(.~diagnosis) +
  ylab('Consensus score')+
  xlab('FDR-adjusted p-value in original HMP2 study')+
  theme_minimal() + 
  scale_color_manual(values = studycolors)+
  geom_smooth(method = 'gam')+
  theme(legend.position='none')+
  stat_cor(method = 'spearman',
          label.y = 1)

consensus.hmp2.ab <- HMP2_metabs %>% 
  clean_names %>% 
  pivot_longer(cols=3:10, 
               names_to = c(".value", 'diagnosis'), 
               names_pattern = "(.+)_(.+)") %>% 
  mutate(diagnosis=str_to_upper(diagnosis)) %>% 
  select(-metabolites) %>% 
  full_join(metab_results, 
            by=c('feature'='HMP2_metab', 'diagnosis'))  %>% 
  select(-KEGG, -hgncid, -SMILES)%>% 
  distinct_all %>% 
  ggplot()+
  aes(x=coefficient, y=MWU_delta, color=diagnosis)+
  geom_jitter(alpha=0.7)+
  facet_grid(.~diagnosis) +
  ylab('Bootstrapped differential abundance')+
  xlab('Differential Abundance in original HMP2 study')+
  theme_minimal() + 
  scale_color_manual(values = studycolors)+
  geom_smooth(method = 'gam')+
  theme(legend.position='none')+
  stat_cor(method = 'spearman',
          label.y = 2)

```

## Upset
```{r}
accessory.df2 <- up.df %>% 
  upset_data(colnames(up.df)[5:6]) %>% 
  `[[`('with_sizes')

met.upset.HMP2 <- up.df %>% 
  upset(colnames(.)[5:6], 
      width_ratio = 0.2,
      keep_empty_groups=F,
      name = 'Metabolite relevance',
      themes = upset_modify_themes(
        list(
          'overall_sizes'=theme(axis.text.x=element_text(angle=90),
                                legend.position = 'none')
          )),
      base_annotations=list(
        'Intersection size'= upset_annotate(
          y = '..count..',
          geom = list(
            geom_bar(data=accessory.df2,
              aes(fill=diagnosis),
              position=position_dodge()),
            scale_fill_manual(values=studycolors),
            geom_text(aes(label=..count.., 
                          y=..count..+20, 
                          group=diagnosis), 
                      stat='count', 
                      position = position_dodge(1))
            )
          )
        ),
      set_sizes=list(
         geom_bar(data =up.df %>%
                      select(1, 5:6) %>% 
                      distinct_all() %>% 
                      upset_data(colnames(.)[2:3]) %>% 
                      `[` (c('intersected','matrix_frame')) %>% 
                      purrr::reduce(left_join) %>% 
                      filter(value) %>% 
                      group_by(group) %>% 
                      summarize(n_mets = n_distinct(feature)) %>% 
                      ungroup, 
                  aes(y=n_mets), 
                  stat = 'identity', 
                  width = .4),
         ylab('Metabolites selected \nper study'),
         scale_y_reverse()
      )
    )
```

## Fig S3
```{r}
figS3top<- consensus.hmp2.ab+
  consensus.hmp2.sig

figS3top/met.upset.HMP2 + 
  patchwork::plot_layout(heights=c(1,3))+
  patchwork::plot_annotation(tag_levels = 'a')

ggsave(file.path(basepath,'results/manuscript_files/FigS3.pdf'), device = 'pdf', width = 10, height = 11, units = 'in')
```

# Supplementary Table 2

```{r Sup2, echo=F}
Sup2 <- target_prioritization %>%  
  filter(ranking > quantile(ranking, .75))  %>% 
  filter(pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', similarity > 0.85, similarity >0.95)) %>% 
  group_by(hgncid) %>%
    mutate(
      mHits = n_distinct(HMDB)
      ) %>%
    mutate(
      Metab_hits = paste0(
        unique(Metabolites),
        collapse = "; ")
      ) %>%
    ungroup %>%
    group_by(HMDB) %>%
    mutate(
      gHits = n_distinct(hgncid)
      ) %>%
    mutate(
      gene_hits = paste0(
        unique(hgncid),
        collapse = "; ")
      ) %>%
    ungroup %>% 
    filter(gHits < 16) %>%
  arrange(desc(ranking)) %>%
  distinct_all %>% 
  group_by(Metabolites, diagnosis) %>% 
  summarize(`∆ab` = mean(MWU_delta, na.rm=T),
            ranking = mean(ranking, na.rm=T),
            analogs = paste0(unique(analog), collapse='; '),
            Targets=paste0(unique(hgncid), collapse='; ')
            )

HMP2_important_metabs <- 
  readxl::read_xlsx('tmp/Supplementary Tables 1-28.xlsx', sheet=16) %>%
  filter(`Minimum Q Value` <= 0.05) %>% 
  left_join(metab_results %>% 
              select(Metabolites, HMP2_metab), 
            by = c('Feature'='HMP2_metab')) %>% 
  select(Metabolites, Feature) %>% 
  distinct(Metabolites, .keep_all = T)

tested_metabs <- 
  metab_results %>% 
  filter(HMDB %in% c('HMDB0001488', # Niacin
                     'HMDB0000684', # L-Kynurenine
                     'HMDB0062781', # 2-oxoglutarate
                     'HMDB0001518', # a-CEHC
                     'HMDB0002364') # oleanolic
         ) %>% 
  distinct(Metabolites) %>%
  pull


Sup2 %>% 
  ungroup %>% 
  mutate(Metabolites=
           case_when(Metabolites %in%  tested_metabs & 
                       Metabolites %in% HMP2_important_metabs$Metabolites~ paste0(Metabolites,footnote_marker_number('† ‡', "html")),
                     Metabolites %in% HMP2_important_metabs$Metabolites ~ paste0(Metabolites, footnote_marker_number('†', "html")),
                     Metabolites %in%  tested_metabs ~ paste0(Metabolites, footnote_marker_number('‡', "html")),
                     TRUE~Metabolites)
         ) %>% 
  kable(format = 'latex', escape = F,  digits = 4) %>%
  kable_styling(bootstrap_options = 'condensed') %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = c(1,5,6), valign = "middle")

Sup2 %>% 
  ungroup %>% 
  mutate(Metabolites=
           case_when(Metabolites %in%  tested_metabs & 
                       Metabolites %in% HMP2_important_metabs$Metabolites~ paste0(Metabolites,footnote_marker_number('† ‡')),
                     Metabolites %in% HMP2_important_metabs$Metabolites ~ paste0(Metabolites, footnote_marker_number('†')),
                     Metabolites %in%  tested_metabs ~ paste0(Metabolites, footnote_marker_number('‡', "html")),
                     TRUE~Metabolites)
         ) %>% 
  cSplit('Targets', 'long', sep = '; ') %>% 
  openxlsx::write.xlsx(file.path(basepath,'results/manuscript_files/Supplementary_Table2.xlsx'))


```


# Host transcriptomics
## Upsets
```{r}
up.htx.df <- htx_results %>% 
  group_by(diagnosis) %>% 
  mutate(in_this_study = case_when(padj<0.05~TRUE, TRUE~FALSE)) %>% 
  ungroup %>% 
  full_join(readxl::read_xlsx(file.path(basepath,'tmp/Supplementary Tables 31-33.xlsx'), sheet = 1) %>% 
               mutate(in_HMP2 = case_when(`FDR Pvalue`<0.05~TRUE,TRUE~FALSE)) %>% 
               select(hgncid=Gene, in_HMP2))  %>% 
  filter_at(vars(in_this_study, in_HMP2), any_vars(.!=FALSE)) %>% 
  distinct(hgncid, diagnosis, log2FoldChange, padj, in_this_study, in_HMP2)

htx.upset.plot <- upset(data = up.htx.df ,
                        intersect = c('in_this_study','in_HMP2'), 
                        width_ratio = 0.2,
                        keep_empty_groups=F,
                        name = 'Differentially expressed genes intersections',
                        themes = upset_modify_themes(
                          list(
                            'overall_sizes'=theme(axis.text.x=element_text(angle=90),
                                                  legend.position = 'none')
                          )
                        ),
      base_annotations=list(
        'Intersection size'= upset_annotate(
          y = '..count..',
          geom = list(
            geom_bar(data=up.htx.df%>% 
                       upset_data(c('in_this_study','in_HMP2')) %>% 
                       `[[`('with_sizes')%>% 
                       drop_na(diagnosis) %>% 
                       distinct(hgncid, diagnosis, intersection),
                     aes(fill=diagnosis),
                     position=position_dodge()),
            scale_fill_manual(values=studycolors, guide=FALSE),
            geom_text(aes(label=..count.., 
                          y=..count..+40, 
                          group=diagnosis), 
                      stat='count', 
                      position = position_dodge(1))
          )
        )
      ),
      annotations = list(
        '-Log(q-value)\ndistribution' = list(
          aes = aes(x=intersection, y = -log(padj), color=diagnosis, size=log2FoldChange),
          geom = list(geom_point(position=position_jitterdodge(),
                                 alpha=0.7),
                      geom_boxplot(alpha=0.8),
                      scale_color_manual(values=studycolors, guide=FALSE),
                      stat_compare_means(label = 'p.signif',
                                         hide.ns = T),
                      guides(size=FALSE)
          )
        )
      ),
      set_sizes=list(
        geom_bar(data =up.htx.df %>%
                   select(1, c('in_this_study','in_HMP2')) %>% 
                   distinct_all() %>% 
                   upset_data(c('in_this_study','in_HMP2')) %>% 
                   `[` (c('intersected','matrix_frame')) %>% 
                   purrr::reduce(left_join) %>% 
                   filter(value) %>% 
                   group_by(group) %>% 
                   summarize(n_targs = n_distinct(hgncid)) %>% 
                   ungroup , 
                 aes(y=n_targs), 
                 stat = 'identity', 
                 width = .4),
        ylab('DEGs selected \nper study'),
        scale_y_reverse()
      )
) 


```

## Pathways

```{r}
pwy.plot <- htx_results %>% 
  distinct(Pathway, highest_p, Fold_Enrichment, diagnosis) %>% 
  arrange(desc(highest_p)) %>% 
  group_by(diagnosis) %>% 
  top_n(-20, highest_p) %>% 
  ggplot()+
  aes(x=reorder(Pathway, -log(highest_p)), y=-log(highest_p))+
  # facet_grid(~diagnosis, scales='free')+
  geom_point(aes(size=log(Fold_Enrichment), color=diagnosis), alpha=.75)+
  coord_flip()+ 
  theme_minimal() + 
  theme(panel.border=element_blank(), 
        legend.background=element_blank())+
  guides(size=guide_legend(title = bquote(~Log[2]~ 'FC')))+
  scale_color_manual(values = studycolors)+
  ylab(bquote(~-Log[10]~adjusted~italic(p)~'-value'))+
  xlab('')

```

## FigS4
```{r}
(htx.upset.plot|pwy.plot) + 
  patchwork::plot_layout(widths=c(5,5))+
  patchwork::plot_annotation(tag_levels = 'a')

ggsave(file.path(basepath,'results/manuscript_files/FigS4.pdf'), device = 'pdf', width = 15, height = 9, units = 'in')

```

# Supplementary Table 3

```{r Sup3}
HMP2_important_htx <- 
  readxl::read_xlsx(file.path(basepath,'tmp/Supplementary Tables 31-33.xlsx'), sheet = 1) %>% 
  group_by(Comparison) %>% 
  filter(`FDR Pvalue`<0.05) %>% 
  ungroup %>% 
  distinct(Gene) %>% 
  pull 

Sup3 <- target_prioritization %>%  
  filter(ranking > quantile(ranking, .75) | !is.na(log2FoldChange),
         pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', similarity > 0.85, similarity >0.95)) %>% 
  group_by(hgncid) %>%
    mutate(
      mHits = n_distinct(HMDB)
      ) %>%
    mutate(
      Metab_hits = paste0(
        unique(Metabolites),
        collapse = "; ")
      ) %>%
    ungroup %>%
    group_by(Metabolites) %>%
    mutate(
      gHits = n_distinct(hgncid)
      ) %>%
    mutate(
      gene_hits = paste0(
        unique(hgncid),
        collapse = "; ")
      ) %>%
    ungroup %>% 
    filter(gHits <= 20, 
           mHits <= 20) %>%
  arrange(desc(ranking)) %>%
  distinct_all %>% 
  select(Targets=hgncid, Metabolites, MWU_delta, analog,
         pxc50, similarity, similarity_type, type,
         log2FoldChange, padj, Pathway, Fold_Enrichment, highest_p, 
         diagnosis) %>% 
  group_by(Metabolites, Targets, type) %>% 
  filter(pxc50 == max(pxc50)) %>% 
  group_by(Targets, Metabolites, diagnosis) %>% 
  drop_na(log2FoldChange) %>% 
  mutate(candidates = 
           case_when(MWU_delta<0 & type=='Negative'~ 'Negative modulator (scarce)', 
                     MWU_delta<0 & type=='Positive'~ 'Positive modulator (scarce)', 
                     MWU_delta>0 & type=='Positive'~ 'Positive modulator (abundant)', 
                     MWU_delta>0 & type=='Negative'~ 'Negative modulator (abundant)', 
                     TRUE~'Unknown')) %>% 
  distinct(Metabolites, pxc50, candidates, Pathway, .keep_all = T) %>% 
  group_by(Targets, diagnosis) %>% 
  summarize(Metabolites = paste0(Metabolites, 
                                 #', Action: ', type 
                                 ', ', candidates,
                                 ' (pxC50=',pxc50,')', collapse='; '),
            `∆expr` = mean(log2FoldChange, na.rm=T), 
            qval = mean(padj, na.rm=T), 
            `Enriched Pathways` = paste0(unique(na.omit(Pathway)), collapse='; '), 
            Metabolites = paste0(unique(Metabolites), collapse= '; ')) %>% 
  distinct_all
  
Sup3 %>% 
  ungroup %>% 
  distinct_all %>% 
  type_convert() %>% 
  mutate(Targets = case_when(Targets %in% HMP2_important_htx ~ paste0(Targets,footnote_marker_number('†', "html")),
                            TRUE~ Targets)) %>% 
  kable(format = 'html', escape=F, digits = 4) %>%
  kable_styling(bootstrap_options = 'condensed') %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = c(1,3:6), valign = "middle")

Sup3 %>% 
  ungroup %>% 
  type_convert() %>% 
  mutate(Targets = case_when(Targets %in% HMP2_important_htx ~ paste0(Targets,footnote_marker_number('†')),
                            TRUE~ Targets))%>% 
  cSplit('Metabolites', 'long', sep = '; ') %>% 
  cSplit('Enriched Pathways', 'long', sep = '; ') %>% 
  distinct_all %>% 
  openxlsx::write.xlsx(file.path(basepath,'results/manuscript_files/Supplementary_Table3.xlsx'))

```
# Supplementary Table 4

```{r Sup4}
Sup4 <-   omim.table%>% 
  distinct(hgncid) %>% 
  collect %>% 
  full_join(IBD_Gwases) %>% 
  full_join(IBD_pathways) %>% 
  inner_join(target_prioritization) %>%
  filter(pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', 
                similarity > 0.85, 
                similarity >0.9)
  ) %>%  
  group_by(hgncid, Metabolites, diagnosis) %>% 
  mutate(candidates = 
           case_when(MWU_delta<0 & type=='Negative'~ 'Negative modulator (scarce)', 
                     MWU_delta<0 & type=='Positive'~ 'Positive modulator (scarce)', 
                     MWU_delta>0 & type=='Positive'~ 'Positive modulator (abundant)', 
                     MWU_delta>0 & type=='Negative'~ 'Negative modulator (abundant)', 
                     TRUE~'Unknown'))  %>% 
  # filter(candidates != 'Unknown') %>%
  group_by(hgncid) %>% 
  mutate(
    mHits = n_distinct(HMDB)
  ) %>%
  ungroup %>% 
  group_by(HMDB) %>% 
  mutate(
    gHits = n_distinct(hgncid)
  ) %>%
  group_by(hgncid, diagnosis) %>% 
  distinct(SNPS, Metabolites, MWU_delta, ranking, .keep_all = T) %>% 
  mutate(Metabolites = case_when(Metabolites %in% Sup2$Metabolites ~ paste0(Metabolites,footnote_marker_number('◊', "html")),
                            TRUE~ Metabolites),
         GWAS = paste0(unique(DISEASE.TRAIT), collapse = '; '),
         variant=paste0(unique(paste0(SNPS, ' (',CONTEXT,')')), collapse = '; '),
         `∆expr` = mean(log2FoldChange, na.rm=T), 
         qval = mean(padj, na.rm=T), 
         `Enriched Pathways` = paste0(unique(Pathway), collapse='; '), 
         Metabolites = paste0(unique(paste0(Metabolites, 
                              #', ∆ab=', round(MWU_delta, 3), ', ranking=', round(ranking, 3),
                              ', ', candidates, 
                              ', (pxc50 = ', pxc50, ')')), 
                              collapse='; ')
  ) 


Sup4 %>% 
  ungroup %>% 
  type_convert() %>% 
  arrange(hgncid, Metabolites) %>% 
  mutate(GWAS = str_remove_all(GWAS, 'GSK500kV3_|gc_|GSK500KV3lin_M2W_|delange17_'), 
         variant = str_replace_all(variant,'_', ' ')) %>%
  select(hgncid, variant, GWAS, `Additional pathway evidence`=action,
         `∆expr`, qval, `Enriched Pathways`, diagnosis, Metabolites) %>% 
  mutate(hgncid = case_when(hgncid %in% Sup3$Targets ~ paste0(hgncid,footnote_marker_number('§')),
                            TRUE~ hgncid)) %>%
  dplyr::rename(Target=hgncid) %>% 
  cSplit('Metabolites', 'long', sep = '; ') %>% 
  cSplit('Enriched Pathways', 'long', sep = '; ') %>% 
  distinct_all() %>% 
  openxlsx::write.xlsx(file.path(basepath,'results/manuscript_files/Supplementary_Table4.xlsx'))
```

# Table 2

```{r}
Sup4 %>% 
  ungroup %>% 
  type_convert() %>% 
  filter(!grepl('Unknown',Metabolites)) %>% 
  arrange(hgncid, Metabolites) %>% 
  mutate(GWAS = str_remove_all(GWAS, 'GSK500kV3_|gc_|GSK500KV3lin_M2W_|delange17_'), 
         variant = str_replace_all(variant,'_', ' ')) %>%
  select(hgncid, variant, GWAS, `Additional pathway evidence`=action,
         `∆expr`, qval, `Enriched Pathways`, diagnosis, Metabolites) %>% 
  mutate(hgncid = case_when(hgncid %in% Sup3$Targets ~ paste0(hgncid,footnote_marker_number('§', "html")),
                            TRUE~ hgncid)) %>%
  dplyr::rename(Target=hgncid) %>% 
  distinct_all() %>%
  kable(format = 'html', escape=F, digits = 3) %>% 
  kable_styling(bootstrap_options = 'condensed') %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = c(1:4,9), valign = "middle", row_group_label_position = 'stack')%>%
  add_header_above(c(" " = 1, "Genetic association" = 3, "Transcriptomic signature" = 4, " "=1), align = 'left')

```

# Supplementary Table 6
```{r}
target_prioritization %>% 
  left_join(omim.table%>% 
    distinct(hgncid) %>% 
    full_join(IBD_Gwases) %>% 
    full_join(IBD_pathways)) %>% 
  filter(pxc50 > 5.5,
         ifelse(similarity_type == 'Tanimoto', similarity > 0.85, similarity >0.95)) %>% 
  group_by(hgncid) %>%
    mutate(
      mHits = n_distinct(Metabolites)
      ) %>%
    mutate(
      Metab_hits = paste0(
        unique(Metabolites),
        collapse = "; ")
      ) %>%
    ungroup %>%
    group_by(Metabolites) %>%
    mutate(
      gHits = n_distinct(hgncid)
      ) %>%
    mutate(
      gene_hits = paste0(
        unique(hgncid),
        collapse = "; ")
      ) %>%
    ungroup %>% 
  select(Metabolites, HMDB, diagnosis, 
         `Metabolite differential concentration`=MWU_delta, 
         `Metabolite scoring` = ranking,
         analog, `Similarity index` = similarity_type, similarity, 
         Modulation=type, pxc50, 
         Target=hgncid, 
         `Target differential expression` = log2FoldChange,
         `Adjusted p-value` = padj, 
         `In enriched pathway` = Pathway, 
         `Enrichment significance` = lowest_p, 
         `Metabolite pleiotropy` = gHits, 
         `Target pleiotropy` = mHits, 
         `Genetic association` = DISEASE.TRAIT,
         Variant = SNPS, 
         `GWAS p-value`=P.VALUE, PUBMEDID, 
         `Manual pathway curation`=action) %>% 
  distinct_all  %>% 
  openxlsx::write.xlsx(file.path(basepath,'results/manuscript_files/Supplementary_Table6.xlsx'))
  
```


```{r}
save.image(file.path('tmp/host_transcriptomics_w_figures.RData'))
rm(list = ls())
sessionInfo()
```



